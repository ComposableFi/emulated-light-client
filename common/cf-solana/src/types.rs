use bytemuck::TransparentWrapper;
use lib::hash::CryptoHash;

/// Cryptographically secure hash.
///
/// The type is completely algorithm agnostic.  It does not provide any methods
/// for computing the hash and in particular may store SHA2 or Blake3 digest.
/// This is in contrast to `solana_program::hash::Hash` or
/// `solana_program::blake3::Hash` which are more strongly typed.
///
/// We’re using a separate type rather than [`CryptoHash`] because we want to
/// have convenient conversion function between our type and Solana’s types.
/// Another aspect is that [`CryptoHash`] is (so far) only used for SHA2 hashes
/// while this type is also used for Blake3 hashes.  This is something that we
/// may end up revisiting.
#[derive(
    Clone,
    Copy,
    Default,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    bytemuck::TransparentWrapper,
    bytemuck::Pod,
    bytemuck::Zeroable,
    derive_more::AsRef,
    derive_more::AsMut,
    derive_more::From,
    derive_more::Into,
)]
#[into(owned, ref, ref_mut)]
#[repr(transparent)]
pub struct Hash(pub [u8; 32]);

/// Solana public key also used as account address.
#[derive(
    Clone,
    Copy,
    Default,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    bytemuck::TransparentWrapper,
    bytemuck::Pod,
    bytemuck::Zeroable,
    derive_more::AsRef,
    derive_more::AsMut,
    derive_more::From,
    derive_more::Into,
)]
#[into(owned, ref, ref_mut)]
#[repr(transparent)]
pub struct PubKey(pub [u8; 32]);


impl From<CryptoHash> for Hash {
    fn from(hash: CryptoHash) -> Hash { Hash(hash.into()) }
}

impl<'a> From<&'a CryptoHash> for &'a Hash {
    fn from(hash: &'a CryptoHash) -> &'a Hash {
        Hash::wrap_ref(hash.as_array())
    }
}

impl From<Hash> for CryptoHash {
    fn from(hash: Hash) -> Self { Self(hash.0) }
}

impl<'a> From<&'a Hash> for &'a CryptoHash {
    fn from(hash: &'a Hash) -> Self { Self::from(&hash.0) }
}


impl From<::blake3::Hash> for Hash {
    fn from(hash: ::blake3::Hash) -> Hash { Hash(hash.into()) }
}

impl<'a> From<&'a ::blake3::Hash> for &'a Hash {
    fn from(hash: &'a ::blake3::Hash) -> &'a Hash {
        Hash::wrap_ref(hash.as_bytes())
    }
}

impl From<Hash> for ::blake3::Hash {
    fn from(hash: Hash) -> Self { Self::from(hash.0) }
}


macro_rules! impl_ref_conversion {
    ($ty:ty) => {
        impl<'a> From<&'a [u8; 32]> for &'a $ty {
            fn from(bytes: &'a [u8; 32]) -> Self { <$ty>::wrap_ref(bytes) }
        }

        impl<'a> From<&'a mut [u8; 32]> for &'a mut $ty {
            fn from(bytes: &'a mut [u8; 32]) -> Self { <$ty>::wrap_mut(bytes) }
        }

        impl<'a> TryFrom<&'a [u8]> for &'a $ty {
            type Error = core::array::TryFromSliceError;

            #[inline]
            fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
                <&[u8; 32]>::try_from(bytes).map(Into::into)
            }
        }

        impl<'a> TryFrom<&'a [u8]> for $ty {
            type Error = core::array::TryFromSliceError;

            #[inline]
            fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
                <&Self>::try_from(bytes).map(Clone::clone)
            }
        }
    };
}

impl_ref_conversion!(Hash);
impl_ref_conversion!(PubKey);


#[allow(unused_macros)]
macro_rules! impl_sol_conversions {
    ($crt:ident) => {
        // ========== $crt::hash::Hash ==========

        impl From<$crt::hash::Hash> for Hash {
            fn from(obj: $crt::hash::Hash) -> Hash { obj.to_bytes().into() }
        }

        impl<'a> From<&'a $crt::hash::Hash> for &'a Hash {
            fn from(obj: &'a $crt::hash::Hash) -> &'a Hash {
                <&[u8; 32]>::try_from(obj.as_ref()).unwrap().into()
            }
        }

        impl From<Hash> for $crt::hash::Hash {
            fn from(obj: Hash) -> Self { Self::from(obj.0) }
        }

        impl<'a> From<&'a Hash> for &'a $crt::hash::Hash {
            fn from(obj: &'a Hash) -> Self {
                let obj = &obj.0 as *const [u8; 32] as *const $crt::hash::Hash;
                // SAFETY: $crt::hash::Hash is repr(transparent)
                unsafe { &*obj }
            }
        }

        // ========== $crt::blake3::Hash ==========

        impl From<$crt::blake3::Hash> for Hash {
            fn from(hash: $crt::blake3::Hash) -> Hash { hash.to_bytes().into() }
        }

        impl<'a> From<&'a $crt::blake3::Hash> for &'a Hash {
            fn from(hash: &'a $crt::blake3::Hash) -> &'a Hash {
                (&hash.0).into()
            }
        }

        impl<'a> From<&'a mut $crt::blake3::Hash> for &'a mut Hash {
            fn from(hash: &'a mut $crt::blake3::Hash) -> &'a mut Hash {
                (&mut hash.0).into()
            }
        }

        impl From<Hash> for $crt::blake3::Hash {
            fn from(hash: Hash) -> Self { Self(hash.0) }
        }

        impl<'a> From<&'a Hash> for &'a $crt::blake3::Hash {
            fn from(hash: &'a Hash) -> Self {
                let hash =
                    &hash.0 as *const [u8; 32] as *const $crt::blake3::Hash;
                // SAFETY: $crt::hash::Hash is repr(transparent)
                unsafe { &*hash }
            }
        }

        impl<'a> From<&'a mut Hash> for &'a mut $crt::blake3::Hash {
            fn from(hash: &'a mut Hash) -> Self {
                let hash =
                    &mut hash.0 as *mut [u8; 32] as *mut $crt::blake3::Hash;
                // SAFETY: $crt::hash::Hash is repr(transparent)
                unsafe { &mut *hash }
            }
        }

        // ========== $crt::pubkey::Pubkey ==========

        impl From<$crt::pubkey::Pubkey> for PubKey {
            fn from(obj: $crt::pubkey::Pubkey) -> PubKey {
                obj.to_bytes().into()
            }
        }

        impl<'a> From<&'a $crt::pubkey::Pubkey> for &'a PubKey {
            fn from(obj: &'a $crt::pubkey::Pubkey) -> &'a PubKey {
                <&[u8; 32]>::try_from(obj.as_ref()).unwrap().into()
            }
        }

        impl<'a> From<&'a mut $crt::pubkey::Pubkey> for &'a mut PubKey {
            fn from(pk: &'a mut $crt::pubkey::Pubkey) -> &'a mut PubKey {
                <&mut [u8; 32]>::try_from(pk.as_mut()).unwrap().into()
            }
        }

        impl From<PubKey> for $crt::pubkey::Pubkey {
            fn from(obj: PubKey) -> Self { Self::from(obj.0) }
        }

        impl<'a> From<&'a PubKey> for &'a $crt::pubkey::Pubkey {
            fn from(obj: &'a PubKey) -> Self {
                let obj =
                    &obj.0 as *const [u8; 32] as *const $crt::pubkey::Pubkey;
                // SAFETY: $crt::pubkey::Pubkey is repr(transparent)
                unsafe { &*obj }
            }
        }

        impl<'a> From<&'a mut PubKey> for &'a mut $crt::pubkey::Pubkey {
            fn from(pk: &'a mut PubKey) -> Self {
                let pk =
                    &mut pk.0 as *mut [u8; 32] as *mut $crt::pubkey::Pubkey;
                // SAFETY: $crt::pk::PubKey is repr(transparent)
                unsafe { &mut *pk }
            }
        }
    };
}

#[cfg(feature = "solana-program")]
impl_sol_conversions!(solana_program);
#[cfg(any(test, feature = "solana-program-2"))]
impl_sol_conversions!(solana_program_2);
