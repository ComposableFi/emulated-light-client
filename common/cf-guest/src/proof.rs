use alloc::string::{String, ToString};
use alloc::vec::Vec;

use guestchain::BlockHeader;
use lib::hash::CryptoHash;

mod ibc {
    pub use ibc_core_commitment_types::commitment::{
        CommitmentPrefix, CommitmentProofBytes, CommitmentRoot,
    };
    pub use ibc_core_host::types::path;
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct IbcProof {
    /// Serialised proof.
    pub proof: Vec<u8>,
    /// Commitment root.
    pub root: CryptoHash,
    /// Value stored at the path (if it exists).
    pub value: Option<CryptoHash>,
}

impl IbcProof {
    /// Returns commitment prefix to use during verification.
    pub fn prefix(&self) -> ibc::CommitmentPrefix { Default::default() }

    /// Returns commitment root.
    pub fn root(&self) -> ibc::CommitmentRoot { self.root.to_vec().into() }

    /// Consumes object and returns commitment proof.
    pub fn proof(self) -> ibc::CommitmentProofBytes {
        self.proof.try_into().unwrap()
    }
}


#[derive(Clone, Debug, PartialEq, Eq, derive_more::From)]
pub enum GenerateError {
    /// State root in block header and root of trie don’t match.
    WrongState,

    /// Error reading data from the trie.
    BadTrie(sealable_trie::Error),

    /// Invalid path.
    BadPath(trie_ids::path_info::Error),
}

/// Generates a proof for given path.
///
/// `block_header` is header whose hash will be the commitment root.  It’s
/// state root must correspond to `trie`’s root.  `path` specifies IBC path
/// of the value that needs proof.
///
/// # Proof format
///
/// In most cases, proof is Borsh-serialised `(guestchain::BlockHeader,
/// sealable_trie::proof::Proof)` pair.  The header at the front is necessary to
/// determine state root (recall that `root` is the block hash and not state
/// root).
///
/// However, if `path` is one of `SeqSend`, `SeqRecv` or `SeqAck` than proof
/// further contain two big-endian encoded `u64` numbers holding the other
/// two sequence numbers.
///
/// For example, if `path` is `SeqRecv`, the `proof` must at the end include
/// send sequence number and ack sequence number.  For example, if next send
/// sequence is `7`, next ack sequence is `5` and path is `SeqRecv` the
/// proof will end with `be(7) || be(5)` (where `be` denotes encoding 64-bit
/// number as big endian).
///
/// This addition is necessary because sequence numbers are stored together
/// within a single trie value.  For example, proving the next receive
/// sequence is `4` requires proving `be(7), be(4), be(5), be(0)].  For
/// verifier to know what value it checks, it needs to be provided all of
/// the sequence numbers.
///
/// (Note that Borsh uses little endian to encode integers so the sequence
/// numbers cannot be simply borsh deserialised.)
pub fn generate<A: sealable_trie::Allocator>(
    block_header: &BlockHeader,
    trie: &sealable_trie::Trie<A>,
    path: ibc::path::Path,
) -> Result<IbcProof, GenerateError> {
    if trie.hash() != &block_header.state_root {
        return Err(GenerateError::WrongState);
    }
    let root = block_header.calc_hash();

    let trie_ids::PathInfo { key, seq_kind, .. } = path.try_into()?;
    let (value, proof) = trie.prove(&key)?;
    let mut proof = borsh::to_vec(&(&block_header, &proof)).unwrap();

    if let Some((value, seq_kind)) = value.as_ref().zip(seq_kind) {
        proof.reserve(16);
        for (idx, val) in value.as_array().chunks_exact(8).take(3).enumerate() {
            if idx != seq_kind as usize {
                proof.extend_from_slice(val);
            }
        }
    }

    Ok(IbcProof { proof, root, value })
}


#[derive(
    Clone, Debug, PartialEq, Eq, derive_more::From, derive_more::Display,
)]
pub enum VerifyError {
    /// Invalid commitment prefix (expected empty).
    BadPrefix,

    /// Invalid commitment root (expected 32 bytes).
    BadRoot,

    /// Invalid path.
    BadPath(trie_ids::path_info::Error),

    /// Failed deserialising the proof.
    ProofDecodingFailure(String),

    /// Invalid sequence value.
    ///
    /// When verifying `SeqSend`, `SeqRecv` and `SeqAck` paths, the `value` to
    /// verify must be `google.protobuf.UInt64Value` holding the sequence
    /// number.  This error indicates that decoding that protocol message
    /// failed.
    WrongSequenceNumber(prost::DecodeError),

    /// Proof verification failed.
    VerificationFailed,
}

impl From<borsh::maybestd::io::Error> for VerifyError {
    fn from(err: borsh::maybestd::io::Error) -> Self {
        Self::ProofDecodingFailure(err.to_string())
    }
}

/// Verifies a proof for given entry or lack of entry.
///
/// `prefix` must be empty, `proof` and `root` must follow format described in
/// [`generate`] function.  `path` indicates IBC path the proof is for and
/// `value` determines value or lack thereof expected at the path.
///
/// # Value hash
///
/// Since sealable trie doesn’t store values but only hashes, when verifying
/// membership proofs the value needs to be converted into a hash.  There are
/// three cases:
///
/// 1. If `path` includes client id, the hash of the value is calculated with
///    the client id mixed in; see [`super::digest_with_client_id`] function.
///
/// 2. If `path` is `SeqSend`, `SeqRecv` or `SeqAck`, the `value` must be
///    `google.protobuf.UInt64Value` protobuf and hash is calculated as
///    concatenation of the three sequence numbers as described in [`generate`].
///
/// 3. Otherwise, the value is simply hashed.
pub fn verify(
    prefix: &[u8],
    mut proof_bytes: &[u8],
    root: &[u8],
    path: ibc::path::Path,
    value: Option<&[u8]>,
) -> Result<(), VerifyError> {
    if !prefix.is_empty() {
        return Err(VerifyError::BadPrefix);
    }
    let root =
        <&CryptoHash>::try_from(root).map_err(|_| VerifyError::BadRoot)?;
    let path = trie_ids::PathInfo::try_from(path)?;

    let (state_root, proof) = {
        let (header, proof): (BlockHeader, sealable_trie::proof::Proof) =
            borsh::BorshDeserialize::deserialize_reader(&mut proof_bytes)?;
        if root != &header.calc_hash() {
            return Err(VerifyError::VerificationFailed);
        }
        (header.state_root, proof)
    };

    let value = if let Some(value) = value {
        Some(if let Some(seq_kind) = path.seq_kind {
            debug_assert!(path.client_id.is_none());
            // If path.seq_kind is set, `value` must be encoded
            // `google.protobuf.UInt64Value` holding the sequence number.
            let seq = <u64 as prost::Message>::decode(value)?.to_be_bytes();

            // Proof is followed by two more sequence numbers this time in
            // big-endian.  We’re keeping sequence numbers together and we
            // need all of them to figure out the hash kept in the trie.
            let (head, tail) = stdx::split_at::<16, u8>(proof_bytes)
                .ok_or_else(|| {
                    VerifyError::ProofDecodingFailure(
                        "Missing sequences".into(),
                    )
                })?;
            let (a, b) = stdx::split_array_ref(head);
            proof_bytes = tail;

            let hash = match seq_kind as u8 {
                0 => [seq, *a, *b, [0u8; 8]],
                1 => [*a, seq, *b, [0u8; 8]],
                2 => [*a, *b, seq, [0u8; 8]],
                _ => unreachable!(),
            };
            CryptoHash(bytemuck::must_cast(hash))
        } else if let Some(id) = path.client_id.as_ref() {
            // If path includes client id, hash stored in the trie is calculated
            // with the id mixed in.
            super::digest_with_client_id(id, value)
        } else {
            // Otherwise, simply hash the value.
            CryptoHash::digest(value)
        })
    } else {
        None
    };

    if !proof_bytes.is_empty() {
        Err(VerifyError::ProofDecodingFailure("Spurious bytes".into()))
    } else if proof.verify(&state_root, &path.key, value.as_ref()) {
        Ok(())
    } else {
        Err(VerifyError::VerificationFailed)
    }
}


#[test]
fn test_proofs() {
    use core::str::FromStr;

    use ibc_core_host::types::identifiers;
    use sealable_trie::nodes::RawNode;

    struct Trie {
        trie: sealable_trie::Trie<memory::test_utils::TestAllocator<RawNode>>,
        header: BlockHeader,
    }

    impl Trie {
        fn set(&mut self, key: &[u8], value: CryptoHash) {
            self.trie.set(key, &value).unwrap();
            self.header.state_root = self.trie.hash().clone();
        }

        fn root(&self) -> &[u8] { self.trie.hash().as_slice() }
    }

    fn assert_path_proof(
        path: ibc::path::Path,
        value: &[u8],
        stored_hash: &CryptoHash,
    ) {
        let trie = sealable_trie::Trie::new(
            memory::test_utils::TestAllocator::new(100),
        );
        let mut trie = Trie {
            header: BlockHeader::generate_genesis(
                guestchain::BlockHeight::from(0),
                guestchain::HostHeight::from(42),
                core::num::NonZeroU64::new(24).unwrap(),
                trie.hash().clone(),
                CryptoHash::test(86),
            ),
            trie,
        };

        // First try non-membership proof.
        let proof = generate(&trie.header, &trie.trie, path.clone()).unwrap();
        assert!(proof.value.is_none());
        verify(&[], &proof.proof, proof.root.as_slice(), path.clone(), None)
            .unwrap();

        // Verify non-membership fails if value is inserted.
        let key = trie_ids::PathInfo::try_from(path.clone()).unwrap().key;
        trie.set(&key, stored_hash.clone());

        assert_eq!(
            Err(VerifyError::VerificationFailed),
            verify(&[], &proof.proof, trie.root(), path.clone(), None)
        );

        // Generate membership proof.
        let proof = generate(&trie.header, &trie.trie, path.clone()).unwrap();
        assert_eq!(Some(stored_hash), proof.value.as_ref());
        verify(
            &[],
            &proof.proof,
            proof.root.as_slice(),
            path.clone(),
            Some(value),
        )
        .unwrap();

        // Check invalid membership proofs
        assert_eq!(
            Err(VerifyError::BadPrefix),
            verify(
                &[1u8, 2, 3],
                &proof.proof,
                proof.root.as_slice(),
                path.clone(),
                Some(value),
            )
        );

        assert_eq!(
            Err(VerifyError::BadRoot),
            verify(&[], &proof.proof, &[1u8, 2, 3], path.clone(), Some(value),)
        );

        assert_eq!(
            Err(VerifyError::ProofDecodingFailure(
                "Unexpected length of input".into()
            )),
            verify(
                &[],
                &[0u8, 1, 2, 3],
                proof.root.as_slice(),
                path.clone(),
                Some(value),
            )
        );

        assert_eq!(
            Err(VerifyError::ProofDecodingFailure("Spurious bytes".into())),
            verify(
                &[],
                &[proof.proof.as_slice(), b"\0"].concat(),
                proof.root.as_slice(),
                path.clone(),
                Some(value),
            )
        );

        assert_eq!(
            Err(VerifyError::VerificationFailed),
            verify(
                &[],
                &proof.proof,
                &CryptoHash::test(11).as_slice(),
                path.clone(),
                Some(value),
            )
        );
    }

    let client_id = identifiers::ClientId::from_str("foo-bar-1").unwrap();
    let connection_id = identifiers::ConnectionId::new(4);
    let port_id = identifiers::PortId::transfer();
    let channel_id = identifiers::ChannelId::new(5);
    let sequence = identifiers::Sequence::from(6);

    let value = b"foo";
    let value_hash = CryptoHash::digest(value);
    let cv_hash = super::digest_with_client_id(&client_id, value);

    let seq_value = prost::Message::encode_to_vec(&20u64);
    let seq_hash = |idx: usize| {
        let mut hash = [[0u8; 8]; 4];
        hash[idx] = 20u64.to_be_bytes();
        CryptoHash(bytemuck::must_cast(hash))
    };

    macro_rules! check {
        ($path:expr) => {
            check!($path, value, &value_hash)
        };
        ($path:expr; having client) => {
            check!($path, value, &cv_hash)
        };
        ($path:expr, $value:expr, $hash:expr) => {
            assert_path_proof($path.into(), $value, $hash)
        };
    }

    check!(ibc::path::ClientStatePath(client_id.clone()); having client);
    check!(ibc::path::ClientConsensusStatePath {
        client_id: client_id.clone(),
        revision_number: 2,
        revision_height: 3,
    }; having client);

    check!(ibc::path::ConnectionPath(connection_id));
    check!(ibc::path::ChannelEndPath(port_id.clone(), channel_id.clone()));

    check!(
        ibc::path::SeqSendPath(port_id.clone(), channel_id.clone()),
        seq_value.as_slice(),
        &seq_hash(0)
    );
    check!(
        ibc::path::SeqRecvPath(port_id.clone(), channel_id.clone()),
        seq_value.as_slice(),
        &seq_hash(1)
    );
    check!(
        ibc::path::SeqAckPath(port_id.clone(), channel_id.clone()),
        seq_value.as_slice(),
        &seq_hash(2)
    );

    check!(ibc::path::CommitmentPath {
        port_id: port_id.clone(),
        channel_id: channel_id.clone(),
        sequence,
    });
    check!(ibc::path::AckPath {
        port_id: port_id.clone(),
        channel_id: channel_id.clone(),
        sequence,
    });
    check!(ibc::path::ReceiptPath {
        port_id: port_id.clone(),
        channel_id: channel_id.clone(),
        sequence,
    });
}
