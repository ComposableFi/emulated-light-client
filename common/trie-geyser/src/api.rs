use std::sync::Arc;

use cf_solana::proof;

type Result<T, E = jsonrpc_core::Error> = ::core::result::Result<T, E>;

/// Data provided by the Witnessed Trie Geyser plugin.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct SlotData {
    /// Proof of the accounts delta hash.
    pub delta_hash_proof: proof::DeltaHashProof,
    /// Proof of the witness trie.
    pub witness_proof: proof::AccountProof,
    /// Trie root account.
    pub root_account: proof::AccountHashData,
}

#[jsonrpc_derive::rpc]
pub trait Methods {
    type Metadata;

    #[rpc(meta, name = "listSlots")]
    fn list_slots(&self, meta: Self::Metadata) -> Result<Vec<u64>>;

    #[rpc(meta, name = "getLatestSlotData")]
    fn get_latest_slot_data(
        &self,
        meta: Self::Metadata,
    ) -> Result<Option<(u64, Arc<SlotData>)>>;

    #[rpc(meta, name = "getSlotData")]
    fn get_slot_data(
        &self,
        meta: Self::Metadata,
        slot: u64,
    ) -> Result<Option<Arc<SlotData>>>;
}

#[test]
fn test_slot_data_serialisation() {
    use cf_solana::types::PubKey;
    use lib::hash::CryptoHash;

    let account_hash_data = cf_solana::proof::AccountHashData::new(
        42,
        &PubKey(CryptoHash::test(1).into()),
        false,
        u64::MAX,
        b"foo",
        &PubKey(CryptoHash::test(2).into()),
    );

    let mut proof = proof::MerkleProof::default();
    let level =
        [CryptoHash::test(10), CryptoHash::test(11), CryptoHash::test(12)];
    proof.push_level(&level, 1);

    let data = SlotData {
        delta_hash_proof: proof::DeltaHashProof {
            parent_blockhash: CryptoHash::test(101),
            accounts_delta_hash: CryptoHash::test(102),
            num_sigs: 103,
            blockhash: CryptoHash::test(104),
            epoch_accounts_hash: None,
        },
        witness_proof: proof::AccountProof { account_hash_data, proof },
        root_account: cf_solana::proof::AccountHashData::new(
            42,
            &PubKey(CryptoHash::test(50).into()),
            false,
            u64::MAX,
            b"trie",
            &PubKey(CryptoHash::test(51).into()),
        ),
    };

    if !cfg!(miri) {
        insta::assert_json_snapshot!(data);
    }
    let serialised = serde_json::to_string(&data).unwrap();

    let deserialised = serde_json::from_str(&serialised).unwrap();
    assert_eq!(data, deserialised);
}
