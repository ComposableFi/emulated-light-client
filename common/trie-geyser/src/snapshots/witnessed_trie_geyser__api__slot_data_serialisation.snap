---
source: common/trie-geyser/src/api.rs
expression: data
---
{
  "delta_hash_proof": {
    "parent_blockhash": "AAAAZQAAAGUAAABlAAAAZQAAAGUAAABlAAAAZQAAAGU=",
    "accounts_delta_hash": "AAAAZgAAAGYAAABmAAAAZgAAAGYAAABmAAAAZgAAAGY=",
    "num_sigs": 103,
    "blockhash": "AAAAaAAAAGgAAABoAAAAaAAAAGgAAABoAAAAaAAAAGg="
  },
  "witness_proof": {
    "account_hash_data": {
      "lamports": 42,
      "owner": "1113diW7qC4UejYmrNwaGTZbE5bsokUdUaunfztqD2",
      "data": "Zm9v",
      "key": "1116GS1EfP7xJU6Yhkt9Xv8BTACkcVxFxApaLznfR3"
    },
    "proof": "ARIAAAAKAAAACgAAAAoAAAAKAAAACgAAAAoAAAAKAAAACgAAAAwAAAAMAAAADAAAAAwAAAAMAAAADAAAAAwAAAAM"
  },
  "root_account": {
    "lamports": 42,
    "owner": "1113Gen6tYXzqXgPfZv3gKu6WGy4oJUpSnEWNcugZMs",
    "data": "dHJpZQ==",
    "key": "1113KHVc1Nj4KBQwSRHzFbMf6W3fg7EJ5FpRAHuaPZt"
  }
}
